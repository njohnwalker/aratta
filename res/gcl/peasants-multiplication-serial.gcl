// serial version of peasant's multiplication

@req(x >= 0 & y >= 0)
a, b, res := x, y, 0 ;

// invariant a * b + res == x * y

do @inv()
[] a > 0 ->
     if
         a % 2 == 1 -> res := res + b ;
     fi
     a := a / 2 ;
     b := b * 2 ;
od
@ens(res == x * y)
