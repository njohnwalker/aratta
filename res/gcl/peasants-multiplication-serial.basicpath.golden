Path 1
  ; Assume x >= 0 & y >= 0
  ; Substitute
      @a -> x
      @b -> y
      @res -> 0
  { Satisfies INVARIANT == PLACEHOLDER }
Path 2
  ; Assume INVARIANT == PLACEHOLDER
  [ Branch 1
      [ Branch 1
          ; Assume a > 0
          ; Assume a % 2 == 1
          ; Substitute
              @res -> res + b
          { HOLE }
      ]
      ; Substitute
          @a -> a / 2
      ; Substitute
          @b -> b * 2
      { HOLE }
  ]
  { Satisfies INVARIANT == PLACEHOLDER }
Path 3
  ; Assume INVARIANT == PLACEHOLDER
  ; Assume ~ a > 0
  { Satisfies res == x * y }